File: pulsar/java/src/main/java/com/examples/hello/pulsar/functions/FunctionRunner.java
package com.examples.hello.pulsar.functions;

import com.examples.hello.pulsar.services.PulsarRunner;
import lombok.*;
import lombok.extern.slf4j.Slf4j;
import org.apache.pulsar.client.api.Consumer;
import org.apache.pulsar.client.api.Producer;
import org.apache.pulsar.functions.api.Context;
import org.apache.pulsar.functions.api.Function;

@Slf4j
@Builder
public class FunctionRunner<I, O> extends PulsarRunner implements FunctionRunnable {
  private final Consumer<I> consumer;
  private final Producer<O> producer;
  private final Function<I, O> function;
  private final Context context;

  @SneakyThrows
  @Override
  protected void callOnce() {
    val msg = consumer.receive();
    log.debug("Received msg(time={} id={})", msg.getEventTime(), msg.getMessageId());

    val input = msg.getValue();
    val output = function.process(input, context);
    log.debug("Processed msg(time={} id={}), input={}, output={})", msg.getEventTime(), msg.getMessageId(), input, output);

    val id = producer.send(output);
    log.debug("Sent msg(id={})", id);

    consumer.acknowledge(msg);
  }

  @SneakyThrows
  @Override
  public void close() {
    consumer.close();
    producer.close();
  }
}

File: pulsar/kotlin/src/main/kotlin/com/example/hello/pulsar/functions/FunctionRunner.kt
package com.example.hello.pulsar.functions

import com.example.hello.pulsar.services.PulsarRunner
import org.apache.pulsar.client.api.Consumer
import org.apache.pulsar.client.api.Producer
import org.apache.pulsar.functions.api.Context
import org.slf4j.LoggerFactory

class FunctionRunner<I, O>(
  private val source: Consumer<I>,
  private val func: org.apache.pulsar.functions.api.Function<I, O>,
  private val context: Context,
  private val sink: Producer<O>
) : PulsarRunner(), FunctionRunnable {
  override fun callOnce() {
    val messages = source.batchReceive()
    if (messages.size() == 0) return
    log.debug("Received batch (size={})", messages.size())
    messages.forEach {
      val output = func.process(it.value, context)
      sink.send(output)
    }
    source.acknowledge(messages)
    log.debug("Processed batch (size={})", messages.size())
  }

  override fun close() {
    stop()
    source.close()
    sink.close()
  }

  companion object {
    private val log = LoggerFactory.getLogger(FunctionRunner::class.java)
  }
}
